mitiq/interface/mitiq_pennylane/conversions.py

```diff
@@ -0,0 +1,80 @@
+# Copyright (C) 2021 Unitary Fund
+#
+# This program is free software: you can redistribute it and/or modify
+# it under the terms of the GNU General Public License as published by
+# the Free Software Foundation, either version 3 of the License, or
+# (at your option) any later version.
+#
+# This program is distributed in the hope that it will be useful,
+# but WITHOUT ANY WARRANTY; without even the implied warranty of
+# MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+# GNU General Public License for more details.
+#
+# You should have received a copy of the GNU General Public License
+# along with this program.  If not, see <https://www.gnu.org/licenses/>.
+
+"""Functions to convert between Mitiq's internal circuit representation and
+Pennylane's circuit representation.
+"""
+
+from cirq import Circuit, Moment
+
+from mitiq.interface.mitiq_qiskit import from_qasm as cirq_from_qasm, to_qasm
+from mitiq.utils import _pop_measurements
+
+from pennylane.wires import Wires
+from pennylane.tape import QuantumTape
+from pennylane.operation import Expectation
+
+from pennylane import from_qasm as pennylane_from_qasm
+
+
+class UnsupportedQuantumTapeError(Exception):
+    pass
+
+
+def from_pennylane(tape: QuantumTape) -> Circuit:
+    """Returns a Mitiq circuit equivalent to the input QuantumTape.
+
+    Args:
+        tape: Pennylane QuantumTape to convert to a Mitiq circuit.
+
+    Returns:
+        Mitiq circuit representation equivalent to the input QuantumTape.
+    """
+    try:
```

Comment:
Looking at the reverse conversion, to pennylane, we label pennylane `Wires()` with a `range(len(n))` pattern, which will necessarily pack `[0,...n-1]` for `n` wires. However, all we're requiring here is that `Wires()` are _sortable_, (potentially with gaps,) rather than densely packed from `[0,...n-1]`.

We could check, here, that the wires are sequentially packed:
```
    for i in range(len(wires)):
        if wires[i] != i:
            raise UnsupportedQuantumTapeError(
                f"The wire labels of the tape must contiguously pack 0 "
                f"to n-1, for n wires."
            )
```

Further, the pennylane documentation specifies that `Wires()` is _already_ a sorted iterable collection. So, I think we can actually count on the `Wires()` to be sorted in the first place with, without the `try` block here that attempts to sort, currently in this method.

Reply:
I vaguely remember a test with wires named "a" (str) and 0 (int) which couldn't be sorted.

Reply:
This was added a few commits below keeping the "try/execpt" and the suggestion by @WrathfulSpatula.

link: https://github.com/unitaryfund/mitiq/pull/836/commits/7283e56ce3bbb3a8920f92a3484f3975be55fb16